pkg/test/ginkgo/queue.go

```diff
@@ -12,141 +13,127 @@ import (
+// abortOnFailure returns an abort function and a context that will be cancelled on the abort.
+func abortOnFailure(parentContext context.Context) (testAbortFunc, context.Context) {
+	testCtx, cancelFn := context.WithCancel(parentContext)
+	return func(testRunResult *testRunResultHandle) {
+		if isTestFailed(testRunResult.testState) {
+			cancelFn()
+		}
+	}, testCtx
 }
 
-func (q *parallelByFileTestQueue) Close() {
-	q.lock.Lock()
-	defer q.lock.Unlock()
-	q.queue = nil
-	q.active = make(map[string]struct{})
-	q.cond.Broadcast()
+// queueAllTests writes all the tests to the channel and closes it when all are finished
+// even with buffering, this can take a while since we don't infinitely buffer.
+func queueAllTests(remainingParallelTests chan *testCase, tests []*testCase) {
+	for i := range tests {
+		curr := tests[i]
+		remainingParallelTests <- curr
+	}
+
+	close(remainingParallelTests)
 }
 
-func (q *parallelByFileTestQueue) Take(ctx context.Context, fn TestFunc) bool {
+// runTestsUntilChannelEmpty reads from the channel to consume tests, run them, and return when the channel is closed.
+func runTestsUntilChannelEmpty(ctx context.Context, remainingParallelTests chan *testCase, testSuiteRunner testSuiteRunner) {
 	for {
-		test, ok := q.pop()
-		if !ok {
-			q.cond.Wait()
-			continue
-		}
-		if test == nil {
-			return false
+		select {
+		// if the context is finished, simply return
+		case <-ctx.Done():
+			return
+
+		case test, ok := <-remainingParallelTests:
+			if !ok { // channel closed, then we're done
+				return
+			}
+			// if the context is finished, simply return
+			if ctx.Err() != nil {
+				return
+			}
+			testSuiteRunner.RunOneTest(ctx, test)
 		}
-		defer q.done(test)
-		fn(ctx, test)
-		return true
 	}
 }
 
-func (q *parallelByFileTestQueue) Execute(ctx context.Context, tests []*testCase, parallelism int, testFn TestFunc) {
-	defer q.Close()
+// tests are currently being mutated during the run process.
+func (q *parallelByFileTestQueue) Execute(ctx context.Context, tests []*testCase, parallelism int, testOutput testOutputConfig, maybeAbortOnFailureFn testAbortFunc) {
+	testSuiteProgress := newTestSuiteProgress(len(tests))
+	testSuiteRunner := &testSuiteRunnerImpl{
+		commandContext:        q.commandContext,
+		testOutput:            testOutput,
+		testSuiteProgress:     testSuiteProgress,
+		maybeAbortOnFailureFn: maybeAbortOnFailureFn,
+	}
+
+	execute(ctx, testSuiteRunner, tests, parallelism)
+}
 
+// execute is a convenience for unit testing
+func execute(ctx context.Context, testSuiteRunner testSuiteRunner, tests []*testCase, parallelism int) {
 	if ctx.Err() != nil {
 		return
 	}
 
-	serial, parallel := splitTests(tests, func(t *testCase) bool { return strings.Contains(t.name, "[Serial]") })
+	serial, parallel := splitTests(tests, isSerialTest)
 
-	r := ring.New(len(parallel))
-	for _, test := range parallel {
-		r.Value = test
-		r = r.Next()
-	}
-	q.queue = r
+	remainingParallelTests := make(chan *testCase, 100)
```

Comment:
My assumption is `100` here is the max number of tests that will run in parallel -- do I have that right?

Reply:
You could run any number of tests in parallel. This is replacing the ring list it used earlier.  The channel holds up to 100 test cases but it constantly gets refed by the go routine below (queueAllTests).  Then if our parallelism is say, 30, it launches 30 go routines in the for loop below that each consume test cases from channel.

Reply:
ah, so I thought the for-loop below was launching n (n=`parallelism`) parallel tests at a time but actually it's launching n go funcs which all pull 1 test at a time from the channel which results in running up to n parallel tests at a time.